File: lib/src/gsy/gsy_ui_config.dart
/// Native UI options passed via [CommonVideoPlayerView.creationParams].
///
/// Shared by Android GSY and iOS SGPlayer. Defaults match
/// [StandardGSYVideoPlayer] / [GSYVideoOptionBuilder] where applicable.
class GsyUiConfig {
  const GsyUiConfig({
    this.enableNativeControls = true,
    this.enableNativeControlsFullscreen = true,
    this.videoTitle = '',
    this.previewVttUrl,
    this.showFullscreenButton = true,
    this.showLockButton = true,
    this.showVolumeToolbar = true,
    this.showSettingsButton = true,
    this.pictureInPictureEnabled = true,
    this.rotateViewAuto = true,
    this.rotateWithSystem = true,
    this.lockLand = false,
    this.needOrientationUtils = true,
    this.showFullAnimation = true,
    this.hideVirtualKey = true,
    this.showPauseCover = true,
    this.needShowWifiTip = true,
    this.surfaceErrorPlay = true,
    this.releaseWhenLossAudio = true,
    this.showDragProgressTextOnSeekBar = false,
    this.dismissControlTime = 2500,
    this.seekRatio = 1,
    this.speed = 1,
    this.looping = false,
    this.seekOnStartMs = -1,
    this.cacheWithPlay = true,
    this.startAfterPrepared = true,
    this.autoFullWithSize = false,
    this.fullHideActionBar = true,
    this.fullHideStatusBar = true,
    this.keepLastFrameWhenComplete = false,
    this.coverUrl,
    this.thumbPlay = true,
    this.ijkEnableAccurateSeek = true,
  });

  /// Native chrome + pan gestures (seek / volume / brightness).
  ///
  /// - Android: GSY touch widget (`setIsTouchWiget`).
  /// - iOS: bottom chrome visibility and GSY-style pan gestures.
  final bool enableNativeControls;

  /// Fullscreen pan gestures (Android GSY `setIsTouchWigetFull`).
  /// iOS uses [enableNativeControls] for both embedded and fullscreen.
  final bool enableNativeControlsFullscreen;

  final String videoTitle;
  final String? previewVttUrl;
  final bool showFullscreenButton;
  final bool showLockButton;
  final bool showVolumeToolbar;
  final bool showSettingsButton;
  final bool pictureInPictureEnabled;
  final bool rotateViewAuto;
  final bool rotateWithSystem;
  final bool lockLand;
  final bool needOrientationUtils;
  final bool showFullAnimation;
  final bool hideVirtualKey;
  final bool showPauseCover;
  final bool needShowWifiTip;
  final bool surfaceErrorPlay;
  final bool releaseWhenLossAudio;
  final bool showDragProgressTextOnSeekBar;
  final int dismissControlTime;
  final double seekRatio;
  final double speed;
  final bool looping;
  final int seekOnStartMs;
  final bool cacheWithPlay;
  final bool startAfterPrepared;
  final bool autoFullWithSize;
  final bool fullHideActionBar;
  final bool fullHideStatusBar;

  /// Keep the last rendered frame when playback completes (both platforms).
  ///
  /// Android mirrors GSY `KeepLastFrameVideo`; iOS hides the cover overlay
  /// so the SGPlayer last frame stays visible.
  final bool keepLastFrameWhenComplete;

  /// Cover / poster image URL (http/https or local `file://`).
  ///
  /// Shown before playback and again on complete unless
  /// [keepLastFrameWhenComplete] is true.
  final String? coverUrl;

  /// Tap the cover to start playback (Android GSY `setThumbPlay`).
  final bool thumbPlay;

  /// IJK FFmpeg accurate seek; reduces keyframe snap-back on drag (IJK core only).
  final bool ijkEnableAccurateSeek;

  Map<String, dynamic> toCreationParams() => <String, dynamic>{
        'gsyUi': <String, dynamic>{
          'enableNativeControls': enableNativeControls,
          'enableNativeControlsFullscreen': enableNativeControlsFullscreen,
          'videoTitle': videoTitle,
          if (previewVttUrl != null) 'previewVttUrl': previewVttUrl,
          'showFullscreenButton': showFullscreenButton,
          'showLockButton': showLockButton,
          'showVolumeToolbar': showVolumeToolbar,
          'showSettingsButton': showSettingsButton,
          'pictureInPictureEnabled': pictureInPictureEnabled,
          'rotateViewAuto': rotateViewAuto,
          'rotateWithSystem': rotateWithSystem,
          'lockLand': lockLand,
          'needOrientationUtils': needOrientationUtils,
          'showFullAnimation': showFullAnimation,
          'hideVirtualKey': hideVirtualKey,
          'showPauseCover': showPauseCover,
          'needShowWifiTip': needShowWifiTip,
          'surfaceErrorPlay': surfaceErrorPlay,
          'releaseWhenLossAudio': releaseWhenLossAudio,
          'showDragProgressTextOnSeekBar': showDragProgressTextOnSeekBar,
          'dismissControlTime': dismissControlTime,
          'seekRatio': seekRatio,
          'speed': speed,
          'looping': looping,
          if (seekOnStartMs >= 0) 'seekOnStartMs': seekOnStartMs,
          'cacheWithPlay': cacheWithPlay,
          'startAfterPrepared': startAfterPrepared,
          'autoFullWithSize': autoFullWithSize,
          'fullHideActionBar': fullHideActionBar,
          'fullHideStatusBar': fullHideStatusBar,
          'keepLastFrameWhenComplete': keepLastFrameWhenComplete,
          if (coverUrl != null) 'coverUrl': coverUrl,
          'thumbPlay': thumbPlay,
          'ijkEnableAccurateSeek': ijkEnableAccurateSeek,
        },
      };
}

File: android/src/main/kotlin/com/keepwan/kinetic_player/gsy/GsyUiConfig.kt
package com.keepwan.kinetic_player.gsy

/**
 * Mirrors [com.shuyu.gsyvideoplayer.builder.GSYVideoOptionBuilder] defaults for
 * [com.shuyu.gsyvideoplayer.video.StandardGSYVideoPlayer].
 */
data class GsyUiConfig(
    val enableNativeControls: Boolean = true,
    val enableNativeControlsFullscreen: Boolean = true,
    val videoTitle: String = "",
    val previewVttUrl: String? = null,
    val showFullscreenButton: Boolean = true,
    val showLockButton: Boolean = true,
    val showVolumeToolbar: Boolean = true,
    val showSettingsButton: Boolean = true,
    val pictureInPictureEnabled: Boolean = true,
    val rotateViewAuto: Boolean = true,
    val rotateWithSystem: Boolean = true,
    val lockLand: Boolean = false,
    val needOrientationUtils: Boolean = true,
    val showFullAnimation: Boolean = true,
    val hideVirtualKey: Boolean = true,
    val showPauseCover: Boolean = true,
    val needShowWifiTip: Boolean = true,
    val surfaceErrorPlay: Boolean = true,
    val releaseWhenLossAudio: Boolean = true,
    val showDragProgressTextOnSeekBar: Boolean = false,
    val dismissControlTime: Int = 2500,
    val seekRatio: Float = 1f,
    val speed: Float = 1f,
    val looping: Boolean = false,
    val seekOnStartMs: Long = -1,
    val cacheWithPlay: Boolean = true,
    val startAfterPrepared: Boolean = true,
    val autoFullWithSize: Boolean = false,
    val fullHideActionBar: Boolean = true,
    val fullHideStatusBar: Boolean = true,
) {
    companion object {
        fun fromCreationParams(params: Map<String, Any?>?): GsyUiConfig {
            @Suppress("UNCHECKED_CAST")
            val ui = params?.get("gsyUi") as? Map<String, Any?> ?: emptyMap()
            return GsyUiConfig(
                enableNativeControls = ui["enableNativeControls"] as? Boolean ?: true,
                enableNativeControlsFullscreen =
                    ui["enableNativeControlsFullscreen"] as? Boolean ?: true,
                videoTitle = ui["videoTitle"] as? String ?: "",
                previewVttUrl = ui["previewVttUrl"] as? String,
                showFullscreenButton = ui["showFullscreenButton"] as? Boolean ?: true,
                showLockButton = ui["showLockButton"] as? Boolean ?: true,
                showVolumeToolbar =
                    ui["showVolumeToolbar"] as? Boolean
                        ?: params?.get("showVolumeToolbar") as? Boolean ?: true,
                showSettingsButton =
                    ui["showSettingsButton"] as? Boolean
                        ?: params?.get("showSettingsButton") as? Boolean ?: true,
                pictureInPictureEnabled =
                    ui["pictureInPictureEnabled"] as? Boolean
                        ?: params?.get("pictureInPictureEnabled") as? Boolean ?: true,
                rotateViewAuto = ui["rotateViewAuto"] as? Boolean ?: true,
                rotateWithSystem = ui["rotateWithSystem"] as? Boolean ?: true,
                lockLand = ui["lockLand"] as? Boolean ?: false,
                needOrientationUtils = ui["needOrientationUtils"] as? Boolean ?: true,
                showFullAnimation = ui["showFullAnimation"] as? Boolean ?: true,
                hideVirtualKey = ui["hideVirtualKey"] as? Boolean ?: true,
                showPauseCover = ui["showPauseCover"] as? Boolean ?: true,
                needShowWifiTip = ui["needShowWifiTip"] as? Boolean ?: true,
                surfaceErrorPlay = ui["surfaceErrorPlay"] as? Boolean ?: true,
                releaseWhenLossAudio = ui["releaseWhenLossAudio"] as? Boolean ?: true,
                showDragProgressTextOnSeekBar =
                    ui["showDragProgressTextOnSeekBar"] as? Boolean ?: false,
                dismissControlTime = ui["dismissControlTime"] as? Int ?: 2500,
                seekRatio = (ui["seekRatio"] as? Number)?.toFloat() ?: 1f,
                speed = (ui["speed"] as? Number)?.toFloat() ?: 1f,
                looping = ui["looping"] as? Boolean ?: false,
                seekOnStartMs = (ui["seekOnStartMs"] as? Number)?.toLong() ?: -1,
                cacheWithPlay = ui["cacheWithPlay"] as? Boolean ?: true,
                startAfterPrepared = ui["startAfterPrepared"] as? Boolean ?: true,
                autoFullWithSize = ui["autoFullWithSize"] as? Boolean ?: false,
                fullHideActionBar = ui["fullHideActionBar"] as? Boolean ?: true,
                fullHideStatusBar = ui["fullHideStatusBar"] as? Boolean ?: true,
            )
        }
    }
}
